### PR TITLE
Github Interceptor to support Owners File Validation

### PR DESCRIPTION
**Pull request by nagasree9**
# Changes

Github interceptor that blocks a pull request trigger from being executed unless invoked by an owner or with a configurable comment by an owner, for example `/ok_to_test`

Ref: #1488 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#tests) (if functionality changed/added)
- [X] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#docs) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commits)
- [X] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```release-note
Github interceptor blocks a pull request trigger from being executed unless invoked by an owner or with a configurable comment by an owner, for example `/ok_to_test`. This feature can be enabled/disabled by feature flagging it true/false
```

### CODE REVIEW

**Comment by dibyom** [comment]
Might be due to https://github.com/tektoncd/plumbing/pull/1252

**Comment by savitaashture** [comment]
> Might be due to [tektoncd/plumbing#1252](https://github.com/tektoncd/plumbing/pull/1252)

Ah i see
Do we have fix for this already ?

**Comment by dibyom** [comment]
> @dibyom Do we need to revert that pr or does that pr require some changes? Integration tests keep failing at same location.

There is something wrong in how the portforward command is setting up its kubeconfig -- not sure why it only affects when running on kind. I don't have the capacity to debug it this week so I'd suggest reverting the plumbing change for running on kind

**Review by khrm — COMMENTED** [review]
/lgtm

**Review by khrm — COMMENTED** [review]
/lgtm

File: docs/interceptors.md [hunk 1]
```diff
@@ -257,6 +257,55 @@ For more information around adding changed files, see the following examples
 - [github-add-changed-files-pr](../examples/v1beta1/github-add-changed-files-pr)
 - [github-add-changed-files-push-cel](../examples/v1beta1/github-add-changed-files-push-cel)
 
+#### Owners validation for pull requests
+
+The GitHub `Interceptor` supports the ability to halt processing on a pull request if the user is not listed in the [owners](https://www.kubernetes.dev/docs/guide/owners/) file or is not a repository/organization member/owner. This feature can be used to prevent unnecessary execution of a PipelineRun or TaskRun. The GitHub `Interceptor` also supports the ability to trigger a PipelineRun/TaskRun through a comment that contains `/ok-to-test` on a pull request by an owner. 
+
+This feature will also work against private GitHub repositories by supplying a [GitHub personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) in the `personalAccessToken` field.
+
+> NOTE: Owners validation requires (at a minimum) the `pull_request` and `issue_comment` GitHub event types.
```
Comment (dibyom) [inline docs/interceptors.md:266]: minor: not sure I understand the "at a minimum" part - pull_request and issue_comment are the only two valid event types right?

File: docs/triggers-api.md [hunk 2]
```diff
@@ -1,11 +1,3 @@
-<!--
```
Comment (dibyom) [inline docs/triggers-api.md:1]: Was this removal intentional?

File: pkg/apis/triggers/v1beta1/trigger_types.go [hunk 3]
```diff
@@ -168,6 +168,17 @@ type GitHubInterceptor struct {
...
 	EventTypes      []string              `json:"eventTypes,omitempty"`
 	AddChangedFiles GithubAddChangedFiles `json:"addChangedFiles,omitempty"`
+	GithubOwners    GithubOwners          `json:"githubOwners,omitempty"`
+}
+
+type GithubOwners struct {
+	Enabled bool `json:"enabled,omitempty"`
+	// This param/variable is required for private repositories or when either EnableOrgMemberCheck, EnableRepoMemberCheck are set to true
+	PersonalAccessToken *SecretRef `json:"personalAccessToken,omitempty"`
+	// Set this as true to allow org members to submit or comment on PR to proceed
```
Comment (dibyom) [inline pkg/apis/triggers/v1beta1/trigger_types.go:178]: Are these options mutually exclusive? What happens if you set orgMemberCheck to true but repoMember to false? 

An alternative could be field that takes an enum of valid org check types - this will reduce the number of fields. Something like:

```go 
type CheckType string

const(
  OrgMembers CheckType = "orgMembers"
  RepoMembers CheckType = "repoMembers"
  None CheckType = "none"
)

type GithubOwners struct {
  ....
  CheckType CheckType `json:"memberCheck,omitempty"`
}

```

Comment (nagasree9) [reply pkg/apis/triggers/v1beta1/trigger_types.go:178]: They are not mutually exclusive. Either of the orgMemberCheck or repoMemberCheck could be true or false. Check will happen based on which is set to true.

File: docs/interceptors.md (hunk 1 above)
Comment (nagasree9) [reply docs/interceptors.md:266]: For Github Interceptor there are other valid event types like `push`. My intention here is to say that for the Owners validation `pull_request` and `issue_comment` event types are the minimum requirement.

File: docs/triggers-api.md (hunk 2 above)
Comment (nagasree9) [reply docs/triggers-api.md:1]: No, I executed the scripts from /hack to auto-generate these changes

File: pkg/apis/triggers/v1beta1/trigger_types.go (hunk 3 above)
Comment (dibyom) [reply pkg/apis/triggers/v1beta1/trigger_types.go:178]: What happens if we set both to true?

File: docs/interceptors.md (hunk 1 above)
Comment (dibyom) [reply docs/interceptors.md:266]: ok, but currently we only support pull_request and issue_comment? there may be others that we support later (like push?)

Comment (nagasree9) [reply docs/interceptors.md:266]: No, for Owners validation these are the only two it would ever support

File: pkg/apis/triggers/v1beta1/trigger_types.go (hunk 3 above)
Comment (nagasree9) [reply pkg/apis/triggers/v1beta1/trigger_types.go:178]: Then orgMemberCheck takes precedence

File: pkg/apis/triggers/v1beta1/trigger_types.go [hunk 4]
```diff
@@ -168,6 +168,28 @@ type GitHubInterceptor struct {
...
+	// Set the checkType to none if neither of repo members or org members can not submit or comment on PR to proceed
+	None CheckType = "none"
+)
+
+type GithubOwners struct {
+	Enabled bool `json:"enabled,omitempty"`
+	// This param/variable is required for private repos or when checkType is set to orgMembers or repoMembers or both
+	PersonalAccessToken *SecretRef `json:"personalAccessToken,omitempty"`
+	// Set the value to one of the supported values (orgMembers, repoMembers, both, none)
+	CheckType CheckType `json:"memberCheck,omitempty"`
```
Comment (dibyom) [inline pkg/apis/triggers/v1beta1/trigger_types.go:192]: the json tag does not match the struct name i.e. 
```suggestion
	CheckType CheckType `json:"checkType,omitempty"`
```

or alternatively

```suggestion
	MemberCheck CheckType `json:"memberCheck,omitempty"`
```

File: pkg/interceptors/github/github_test.go [hunk 5]
```diff
@@ -1129,3 +1129,917 @@ func Test_makeClient(t *testing.T) {
 		})
 	}
 }
+
+func Test_getPersonalAccessTokenSecret(t *testing.T) {
+
+	ctx, _ := test.SetupFakeContext(t)
+	var secretToken = "secret"
+	// interceptor := &Interceptor{
```
Comment (dibyom) [inline pkg/interceptors/github/github_test.go:1137]: remove instead of commenting out

File: pkg/apis/triggers/v1beta1/trigger_types.go (hunk 4 above)
Comment (khrm) [inline pkg/apis/triggers/v1beta1/trigger_types.go:192]: This should be:

```
	CheckType CheckType `json:"checkType,omitempty"`
```

File: pkg/interceptors/github/github_test.go [hunk 6]
```diff
@@ -1129,3 +1129,917 @@ func Test_makeClient(t *testing.T) {
...
+
+	type args struct {
+		ctx context.Context
+		r   *triggersv1.InterceptorRequest
+		p   triggersv1.GitHubInterceptor
+	}
+	tests := []struct {
+		name   string
+		secret *corev1.Secret
+		// w       *Interceptor
```
Comment (dibyom) [inline pkg/interceptors/github/github_test.go:1149]: same as above

**Review by khrm — COMMENTED** [review]
Can you recheck this example? It's giving error.

**Review by dibyom — COMMENTED** [review]
Almost there - just a couple of minor comments

File: docs/interceptors.md (hunk 1 above)
Comment (dibyom) [reply docs/interceptors.md:266]: I'd just rephrase this to say Owners validation requires the event type to be pull_request or issue_comment - the "at a minimum" phrase I think is a bit confusing

**Review by khrm — COMMENTED** [review]
/kind feature

File: docs/interceptors.md [hunk 7]
```diff
@@ -257,6 +257,54 @@ For more information around adding changed files, see the following examples
...
+              value:
+                secretName: github-secret
+                secretKey: secretToken
+            - name: "eventTypes"
+              # Owners validation requires (at a minimum) the `pull_request` and `issue_comment` GitHub event types
+              value: ["pull_request", "issue_comment"]
+            - name: "githubOwners"
+              value: 
+                enabled: true
+                # This value is needed for private repos or when checkType is set to orgMembers or repoMembers or both
```
Comment (khrm) [inline docs/interceptors.md:297]:
```suggestion
                # This value is needed for private repos or when checkType is set to orgMembers or repoMembers or all
```

**Comment by khrm** [comment]
You might need to run  ./update-codegen.sh after doing these changes.

**Comment by khrm** [comment]
Sorry, by mistake I closed this. I wanted to comment. So please after those fixes, run ./update-codegen.sh . It should fix the error.

**Review by khrm — COMMENTED** [review]
/lgtm
